tensordict/prototype/tensorclass.py

```diff
@@ -481,7 +483,26 @@ def _setitem(self, item: NestedKey, value: Any) -> None:
         isinstance(item, tuple) and all(isinstance(_item, str) for _item in item)
     ):
         raise ValueError("Invalid indexing arguments.")
-    if is_tensorclass(value) and not isinstance(value, self.__class__):
+
+    if not is_tensorclass(value) and not isinstance(
+        value, (TensorDictBase, numbers.Number, Tensor, MemmapTensor)
+    ):
+        raise ValueError(
+            f"__setitem__ only supports tensorclasses, tensordicts,"
+            f" numeric scalars and tensors. Got {type(value)}"
+        )
+
+    # if it is one of accepted "broadcast" types
+    if not is_tensorclass(value):
+        # attempt broadcast on all tensordata and nested tensorclasses
+        self._tensordict[item] = value
+        for key, val in self._non_tensordict.items():
+            if is_tensorclass(val):
+                _setitem(self._non_tensordict[key], item, value)
+        return
```

Comment:
I believe as a result of this change, [this code](https://github.com/pytorch-labs/tensordict/pull/226/files#diff-8327839160f3184ce4a4e04c9e19ec4afbf392c5c3705fe8a3c534162d32e6eeR515-R517) below is now unreachable. Perhaps we can restructure the control flow here?

Reply:
right, done